### PR TITLE
Init implementation of snapshot function

### PR DESCRIPTION
**Pull request by bellofreedom**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,3 @@ target
 .idea
 halodb.iml
 tmp/
-
```
Comment (wangtao724) [inline .gitignore]: revert unnecessary change

File: src/main/java/com/oath/halodb/CompactionManager.java [hunk 2]
```diff
@@ -44,6 +44,7 @@
     private static final int STOP_SIGNAL = -10101;
 
     private final ReentrantLock startStopLock = new ReentrantLock();
+    private final Lock compactionFileRollOverLock = new ReentrantLock();
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/CompactionManager.java]: Any reason to define it as ```Lock``` instead of its actual type ```ReentrantLock```. It not required, please keep consistent to existing code. Then no need to import ```java.util.concurrent.locks.*```

File: src/main/java/com/oath/halodb/CompactionManager.java [hunk 3]
```diff
@@ -279,13 +280,16 @@ private void copyFreshRecordsToNewFile(int idOfFileToCompact) throws IOException
                     else {
                         // write thread wrote a new version while this version was being compacted.
                         // therefore, this version is stale.
+                        // TODO: From the design, currentWriteFile should be be compacted?
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/CompactionManager.java]: Why add this comment? It is no relevant code change, better not introduce this kind of comment.

File: src/main/java/com/oath/halodb/CompactionManager.java [hunk 4]
```diff
@@ -279,13 +280,16 @@ private void copyFreshRecordsToNewFile(int idOfFileToCompact) throws IOException
...
                         dbInternal.addFileToCompactionQueueIfThresholdCrossed(currentWriteFile.getFileId(), recordSize);
                     }
                     currentWriteFileOffset += recordSize;
                     currentWriteFile.setWriteOffset(currentWriteFileOffset);
                 }
             }
 
+
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/CompactionManager.java]: Remove these 2 new lines

File: src/main/java/com/oath/halodb/CompactionManager.java [hunk 5]
```diff
@@ -306,18 +310,33 @@ private boolean isRecordFresh(IndexFileEntry entry, InMemoryIndexMetaData metaDa
         }
 
         private void rollOverCurrentWriteFile(int recordSize) throws IOException {
-            if (currentWriteFile == null ||  currentWriteFileOffset + recordSize > dbInternal.options.getMaxFileSize()) {
-                if (currentWriteFile != null) {
-                    currentWriteFile.flushToDisk();
-                    currentWriteFile.getIndexFile().flushToDisk();
+            // Make sure there's no unnecessary roll over of the write
+            compactionFileRollOverLock.lock();
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/CompactionManager.java]: This is a critical path which each write will go through. This lock may result unnecessary cost. Performance test is needed. Can we find another way to do it more efficiently?

Comment (wangtao724) [reply src/main/java/com/oath/halodb/CompactionManager.java]: Looks like you have used lock in ```forceRolloverCurrentWriteFile```. Is this lock still necessary?

File: src/main/java/com/oath/halodb/CompactionManager.java [hunk 6]
```diff
@@ -306,18 +310,33 @@ private boolean isRecordFresh(IndexFileEntry entry, InMemoryIndexMetaData metaDa
...
+    void forceRolloverCurrentWriteFile() throws IOException {
+        compactionFileRollOverLock.lock();
+        try {
+            if (currentWriteFile != null) {
+                currentWriteFile.flushToDisk();
+                currentWriteFile.getIndexFile().flushToDisk();
+            }
+            currentWriteFile = dbInternal.createHaloDBFile(HaloDBFile.FileType.COMPACTED_FILE);
+            dbInternal.getDbDirectory().syncMetaData();
+            currentWriteFileOffset = 0;
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/CompactionManager.java]: ```currentWriteFileOffset``` is used by compaction thread all the time. When you do the rollover, it will impact ongoing writes. At least this variable need to be ```volatile```.

Considering user not often do snapshot, can we pause compaction job during taking snapshot?

File: src/main/java/com/oath/halodb/FileUtils.java [hunk 7]
```diff
@@ -30,6 +32,17 @@ static void createDirectoryIfNotExists(File directory) throws IOException {
         }
     }
 
+    static void deleteDirectory(File dir) throws IOException {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/FileUtils.java]: ```org.apache.commons.io.FileUtils``` is doing a better job

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 8]
```diff
@@ -281,6 +286,66 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    public boolean takeSnapShot() {
+
+        logger.info("Start generating the snapshot");
+
+        File[] filesToCopy = dbDirectory.getPath().toFile()
+            .listFiles(file -> {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Any reason not copying all files including ```META``` and ```LOCK```. When do partition migration, these files are also needed, right?

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 9]
```diff
@@ -281,6 +286,66 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+                    || Constants.INDEX_FILE_PATTERN.matcher(file.getName()).matches()
+                    || Constants.TOMBSTONE_FILE_PATTERN.matcher(file.getName()).matches());
+            } );
+
+        logger.info("file number need to be copied: {}", filesToCopy.length);
+
+        try {
+            Path dbDirectoryPath = dbDirectory.getPath();
+            File snapShotDir = Paths.get(dbDirectoryPath.toFile().getAbsolutePath(), "snapshot").toFile();
+            // Only support one snapshot now
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: This comment should be moved right above L304

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 10]
```diff
@@ -281,6 +286,66 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    public boolean takeSnapShot() {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: What will happen if multiple threads this method concurrently? Make it ```synchronize```?

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 11]
```diff
@@ -281,6 +286,66 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+        logger.info("file number need to be copied: {}", filesToCopy.length);
+
+        try {
+            Path dbDirectoryPath = dbDirectory.getPath();
+            File snapShotDir = Paths.get(dbDirectoryPath.toFile().getAbsolutePath(), "snapshot").toFile();
+            // Only support one snapshot now
+            // TODO: support multiple snapshots if needed
+            if (snapShotDir.exists()) {
+                logger.warn("The snapshot dir is already existed. Delete the old one.");
+                FileUtils.deleteDirectory(snapShotDir);
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Can't assume ```snapshotDir``` must be a directory. If it is file, it will result exception. Use ```FileUtils .forceDelete()``` instead

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 12]
```diff
@@ -281,6 +286,66 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+                return (Constants.DATA_FILE_PATTERN.matcher(file.getName()).matches()
+                    || Constants.INDEX_FILE_PATTERN.matcher(file.getName()).matches()
+                    || Constants.TOMBSTONE_FILE_PATTERN.matcher(file.getName()).matches());
+            } );
+
+        logger.info("file number need to be copied: {}", filesToCopy.length);
+
+        try {
+            Path dbDirectoryPath = dbDirectory.getPath();
+            File snapShotDir = Paths.get(dbDirectoryPath.toFile().getAbsolutePath(), "snapshot").toFile();
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Minor: snapshotDir

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 13]
```diff
@@ -281,6 +286,66 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+            // TODO: support multiple snapshots if needed
+            if (snapShotDir.exists()) {
+                logger.warn("The snapshot dir is already existed. Delete the old one.");
+                FileUtils.deleteDirectory(snapShotDir);
+            }
+
+            FileUtils.createDirectoryIfNotExists(snapShotDir);
+            logger.info("Created directory for snapshot {}", snapShotDir.toString());
+
+            for (File file : filesToCopy) {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: If ```currentWriteFile``` is rolled over right after the file list finalized, the new file will be forcefully rolled over again.

File: src/main/java/com/oath/halodb/CompactionManager.java (hunk 5 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/CompactionManager.java]: There are three files are supposed to be actively writing when roll over happened. So each file Rollover need a separate Lock for that.

Those Reenentry Locks should be very light weight as it only get blocked when two threads are accessing the Locks. From my test in bucket host, there's no system stunk when rollover happened.

File: src/main/java/com/oath/halodb/CompactionManager.java (hunk 3 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/CompactionManager.java]: Yes, it's another minor issue I found when I adding this feature. I'll address it in another PR.

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 8 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: Existing META would introduce misleading information if adding to the snapshot. And Lock is just some runtime objects which not need to be copied.

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 14]
```diff
@@ -281,6 +286,66 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+            }
+        } catch (IOException e) {
+            logger.warn("IO exception in copy, snapshot failure");
+
+            return false;
+        }
+
+        // TODO: Need solutions to handle the file pieces after the rollover
+        // especially there are frequently rollovers
+        writeLock.lock();
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: It seems that ```writeLock``` is enough for the whole procedure. The workflow should be:
1. Pause compaction
2. writeLock.lock();
3. put current write file to compaction list and then rollover current write file
4. Rollover current tombstone file
5. Construct snapshot file list by **excluding** current write file and tombstone file
6. Put current compaction file to compaction list and rollover compaction file
7.  writeLock.unlock();
8. Create file links for snapshot
9. Resume compaction job

**Comment by wangtao724** [comment]
@bellofreedom I would like to recommend to create snapshot with below steps without introduce too many locks and risk of concurrency. Let's discuss this offline.

It seems that ```writeLock``` is enough for the whole procedure. The workflow should be:
1. Pause compaction
2. writeLock.lock();
3. put current write file to compaction list and then rollover current write file
4. Rollover current tombstone file
5. Construct snapshot file list by **excluding** current write file and tombstone file
6. Put current compaction file to compaction list and rollover compaction file
7.  writeLock.unlock();
8. Create file links for snapshot
9. Resume compaction job

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 14 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: Yes, looks writeLock should be enough for tombstone and currentWriteFile.

File: .gitignore (hunk 1 above)
Comment (bellofreedom) [reply .gitignore]: It maybe due to some IDE auto-edit. I think it's OK to keep this minor change here as we don't need space line in gitignore.

File: src/main/java/com/oath/halodb/FileUtils.java (hunk 7 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/FileUtils.java]: As there's already a class with same name. I don't want to introduce apache FileUtils as it would be somewhat confused.

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 11 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: This function would handle the case if it's a file. File.listFiles() would return null if it's a file type.

File: src/main/java/com/oath/halodb/FileUtils.java (hunk 7 above)
Comment (wangtao724) [reply src/main/java/com/oath/halodb/FileUtils.java]: Then add it to FileUtils and call FileUtils here

Comment (bellofreedom) [reply src/main/java/com/oath/halodb/FileUtils.java]: It's quite a simple function. We could just refer the implementation of the lib.

Comment (wangtao724) [reply src/main/java/com/oath/halodb/FileUtils.java]: Just noticed, it is already in FileUtils. The name ```deleteDirectory``` is not appropriate here because they function with assumption no embedded folder in it. Because it is only used to delete snapshot. How about rename to ```deleteSnapshotDir``` ?

File: src/main/java/com/oath/halodb/HaloDB.java [hunk 15]
```diff
@@ -88,6 +88,10 @@ public void pauseCompaction() throws HaloDBException {
         }
     }
 
+    public void snapshot() {
+        dbInternal.takeSnapShot();
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDB.java]: minor: ```takeSnapshot```. Different meaning between "snap shot" and "snapshot"

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 16]
```diff
@@ -281,6 +285,91 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    public boolean takeSnapShot() {
+        synchronized (snapshotLock) {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: ```synchronized``` method can achieve same impact but clear context if we don't have other methods defined as ```synchronized```

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 17]
```diff
@@ -281,6 +285,91 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+                while (compactionManager.isCompactionRunning()) {
+                    try {
+                        Thread.sleep(1000);
+                    } catch (InterruptedException e) {
+                        logger.info("The snapshot operation was interrupted");
+                        return false;
+                    }
+                }
+
+                File[] filesToCopy = dbDirectory.getPath().toFile()
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: A edge case is current write file or tombstone file are rolled over right after you build up this list.  To be safe, I suggest:
```java
File[] filesToCopy = null; 
writeLock.lock();
try {
    // This is the real cutoff time point
    filesToCopy = dbDirectory.getPath().toFile()..listFiles();
    currentTombstoneFile = forceRollOverTombstoneFile(currentTombstoneFile);
    forceRollOverCurrentWriteFile();
} catch (IOException e) {
    logger.warn("IO exception when rollover current write files", e);
} finally {
    writeLock.unlock();
}

// filter file, create link and so on
...
```

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 18]
```diff
@@ -281,6 +285,91 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    public boolean takeSnapShot() {
+        synchronized (snapshotLock) {
+            logger.info("Start generating the snapshot");
+
+            try {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Before start, check ```isClosing```, ```isOpening``` (to be added) and ```isTombstoneFilesMerging```, if any one is true, should return.

File: src/main/java/com/oath/halodb/HaloDB.java [hunk 19]
```diff
@@ -88,6 +88,10 @@ public void pauseCompaction() throws HaloDBException {
         }
     }
 
+    public void snapshot() {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDB.java]: Need another API to delete snapshot

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 20]
```diff
@@ -281,6 +285,91 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+                    forceRollOverCurrentWriteFile();
+                } catch (IOException e) {
+                    logger.warn("IO exception when rollover current write files", e);
+                } finally {
+                    writeLock.unlock();
+                }
+
+                try {
+                    compactionManager.forceRolloverCurrentWriteFile();
+                } catch (Exception e) {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Should be ```IOException```?

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 21]
```diff
@@ -281,6 +285,91 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+
+            try {
+                try {
+                    compactionManager.pauseCompactionThread();
+                } catch (IOException e) {
+                    logger.warn("IOException when pause the compaction thread", e);
+                    return false;
+                }
+
+                while (compactionManager.isCompactionRunning()) {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Consider merge my PR https://github.com/yahoo/HaloDB/pull/40 first? ;)

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 22]
```diff
@@ -281,6 +285,91 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+                        if (file.exists()) {
+                            Path dest = Paths.get(snapshotDir.getAbsolutePath(), file.getName());
+                            logger.debug("Create links from file {} to {}", file.getName(),
+                                         dest.toFile().getAbsoluteFile());
+                            Files.createLink(dest, file.toPath());
+                        } else {
+                            logger.warn("File {} maybe deleted by compaction job", file.getName());
+                        }
+                    }
+                } catch (IOException e) {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Try to avoid using so many ```try/catch``` for code legibility.  I would suggest simplify to:
```java
try {
    // do the work
    writeLock.lock();
    try { 
     // do the work
    } finally {
        writeLock.unlock();
    }
    // do other work
} catch(IOException e) {
// log error and return false;
} finally {
    compactionManager.resumeCompaction();
}
```

Comment (wangtao724) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: To be consistent with existing public methods, let exception thrown out instead of catching them might be another choice

**Comment by amannaly** [comment]
Correct me if I am wrong: this change is to create a copy of the db as it exists at a particular time without stopping writes? What will this be used for?

**Comment by wangtao724** [comment]
@amannaly Glad see you back and involved. Yes, as what ```snapshot``` itself means, it will take the picture at a specific time point. About "without stoping write", technically, all write are stopped at the time the snapshot is taken (construct the file list), then resume to normal after that. Once we have the file list, the next step is create hard links to the files and put the link file to snapshot folder

**Comment by wangtao724** [comment]
@amannaly About the user cases: we have a daily partition dump job to dump the partition data and do data analysis. Without this feature, the output data are not quite accurate because we can't stop writes for dump job which will run for a few hours.
Another user case is db/partition movement from one server to another. With this feature, we don't need to stop writes during partition copy

**Comment by wangtao724** [comment]
@bellofreedom There are some scenarios need unit test to cover:
1. Interruption during building links, make sure the half done snapshot be cleaned up
2. After snapshot is taken, update all records and wait for all files included in snapshot are compacted. Then verify the files still exist until snapshot is deleted, i.e. files still exist before snapshot deletion. Files got deleted after snapshot deletion

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 16 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: There's another method in using synchronized in class level. It's better to use a separate lock for this function to avoid potential blocking of other methods.

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 17 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: Nice catch. I'll adjust the code.

File: src/main/java/com/oath/halodb/HaloDB.java (hunk 19 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDB.java]: I'm not sure if do we really need the delete function. I'll try to add a new function getSnapshotPath to let the client to make the decision.

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 18 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: Nice points.

**Comment by bellofreedom** [comment]
HI @wangtao724 , 
For the 1st case you mentioned above. There's no mechanism to detect if we have a full copy. If the snapshot was disrupted when creating the links, the clients can't aware the data/files missing. Need to add some manifest/METADATA to point out how many files were created. So when loading the snapshot, we could check the completeness of the snapshot.

I'll address this issue in another PR. Let's keep this PR as the init implementation of snapshot to just include the basic functionality.

File: src/main/java/com/oath/halodb/CompactionManager.java [hunk 23]
```diff
@@ -43,7 +43,7 @@
 
     private static final int STOP_SIGNAL = -10101;
 
-    private final ReentrantLock startStopLock = new ReentrantLock();
+    private final Lock startStopLock = new ReentrantLock();
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/CompactionManager.java]: Better not include irrelevant changes

File: src/main/java/com/oath/halodb/CompactionManager.java [hunk 24]
```diff
@@ -8,14 +8,14 @@
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.util.concurrent.RateLimiter;
 
+import java.util.concurrent.locks.*;
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/CompactionManager.java]: Better not include irrelevant changes

File: src/main/java/com/oath/halodb/CompactionManager.java [hunk 25]
```diff
@@ -8,14 +8,14 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
 import java.nio.channels.FileChannel;
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.LinkedBlockingQueue;
-import java.util.concurrent.locks.ReentrantLock;
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/CompactionManager.java]: Better not include irrelevant changes

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 16 above)
Comment (wangtao724) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: Consider put ```synchronized``` to ```open()```, ```close``` and ```snapshot``` to prevent from concurrent operations. And you can use a separate lock for ```resetStats``` @amannaly Any suggestion?

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 26]
```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
         }
     }
 
+    //TODO: use fine-grained lock if possible
+    public synchronized boolean takeSnapshot() {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: remove ```publish```, this is for internal call

File: src/main/java/com/oath/halodb/FileUtils.java [hunk 27]
```diff
@@ -30,6 +34,21 @@ static void createDirectoryIfNotExists(File directory) throws IOException {
...
+        File[] files = dir.listFiles();
+        if (files != null) {
+            for (File file : files) {
+                if (file.isDirectory()) {
+                    FileUtils.deleteDirectory(file);
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/FileUtils.java]: ```deleteDirectory(file);``` . ```FileUtils``` is not needed. It will make code ```recursion``` legible

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 28]
```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+            writeLock.lock();
+            // TODO: Need solutions to handle the file pieces after the rollover
+            // especially there are frequently rollovers
+            try {
+                forceRollOverCurrentWriteFile();
+                currentTombstoneFile = forceRollOverTombstoneFile(currentTombstoneFile);
+
+                currentWriteFileId = currentWriteFile.getFileId();
+            } catch (IOException e) {
+                logger.warn("IO exception when rollover current write files", e);
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Do we need to catch exception here? Let it thrown out?

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 29]
```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+
+                return false;
+            } finally {
+                writeLock.unlock();
+            }
+
+            File[] filesToCopy = dbDirectory.getPath().toFile()
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: ```filesToLink```?

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 30]
```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+            if (snapshotDir.exists()) {
+                logger.warn("The snapshot dir is already existed. Delete the old one.");
+                FileUtils.deleteDirectory(snapshotDir);
+            }
+
+            FileUtils.createDirectoryIfNotExists(snapshotDir);
+            logger.info("Created directory for snapshot {}", snapshotDir.toString());
+
+            writeLock.lock();
+            // TODO: Need solutions to handle the file pieces after the rollover
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: This may not be needed as long as the percentage of stale data is well tracked. I would suggest to remove this comment. Or it may give user an impression this is a half done work. As long as it works well, TODO == WILL NEVER DO

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 31]
```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+                });
+
+            compactionManager.forceRolloverCurrentWriteFile();
+
+            logger.info("file number need to be copied: {}", filesToCopy.length);
+
+            for (File file : filesToCopy) {
+                if (file.exists()) {
+                    Path dest = Paths.get(snapshotDir.getAbsolutePath(), file.getName());
+                    logger.debug("Create links from file {} to {}", file.getName(),
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: minor: ```Create link from ..```, no 's'

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 32]
```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+                    Matcher m = Constants.STORAGE_FILE_PATTERN.matcher(file.getName());
+                    return  m.matches() && (Integer.parseInt(m.group(1)) < currentWriteFileId);
+                });
+
+            compactionManager.forceRolloverCurrentWriteFile();
+
+            logger.info("file number need to be copied: {}", filesToCopy.length);
+
+            for (File file : filesToCopy) {
+                if (file.exists()) {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Isn't this always true? Do we need this check?

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 33]
```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+                    Path dest = Paths.get(snapshotDir.getAbsolutePath(), file.getName());
+                    logger.debug("Create links from file {} to {}", file.getName(),
+                                 dest.toFile().getAbsoluteFile());
+                    Files.createLink(dest, file.toPath());
+                } else {
+                    logger.warn("There are something wrong when create links about file {}", file.getName());
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: If a supposed existing file doesn't exist, there must be some external interruption. Better to fail snapshot to avoid building a corrupted snapshot

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 34]
```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+
+        return true;
+    }
+
+    public File getSnapshotDirectory() {
+        Path dbDirectoryPath = dbDirectory.getPath();
+        return Paths.get(dbDirectoryPath.toFile().getAbsolutePath(), SNAPSHOT_SUBDIR).toFile();
+    }
+
+    public boolean clearSnapshot() {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Remove public

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 35]
```diff
@@ -281,6 +288,99 @@ int get(byte[] key, ByteBuffer buffer) throws IOException {
...
+
+            return false;
+        } finally {
+            compactionManager.resumeCompaction();
+        }
+
+        return true;
+    }
+
+    public File getSnapshotDirectory() {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Remove public

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 36]
```diff
@@ -323,34 +423,45 @@ private InMemoryIndexMetaData writeRecordToFile(Record record) throws IOExceptio
         return currentWriteFile.writeRecord(record);
     }
 
-    private void rollOverCurrentWriteFile(Record record) throws IOException, HaloDBException {
+    private void rollOverCurrentWriteFile(Record record) throws IOException {
         int size = record.getKey().length + record.getValue().length + Record.Header.HEADER_SIZE;
+        if ((currentWriteFile == null || currentWriteFile.getWriteOffset() + size > options.getMaxFileSize())
+            && !isClosing) {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Why rollover not allowed during closing?  Looks like this is not relevant to snapshot feature. If required, please open  a new PR to fix that.

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 30 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: In fact, the data and datac file would be well maintained. The issue is the tombstone file, it would not be compacted till the restart of the storage engine. That maybe a issue for long run. Looks related to existing issue which the tombstone can't be compacted until the restart.

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 36 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: I just refactored the structure of the old function. There's no logic change in this.

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 32 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: This check is for previous design. Looks no longer needed, I'll remove the check. If there's some issue, let the exception been throwed.

File: src/main/java/com/oath/halodb/HaloDBInternal.java (hunk 28 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/HaloDBInternal.java]: This exception could be handled inside the HaloDB. It would return the false which means the snapshot failed.

File: src/main/java/com/oath/halodb/FileUtils.java (hunk 27 above)
Comment (bellofreedom) [reply src/main/java/com/oath/halodb/FileUtils.java]: In fact calling FileUtils.deleteDirectory(file) and deleteDirectory(file) should be exactly same except the format. 

What do you mean recursion legible?

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 37]
```diff
@@ -35,13 +39,15 @@
 import java.util.concurrent.locks.ReentrantLock;
 
 class HaloDBInternal {
+
     private static final Logger logger = LoggerFactory.getLogger(HaloDBInternal.class);
+    public static final String SNAPSHOT_SUBDIR = "snapshot";
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: remove public

File: src/main/java/com/oath/halodb/HaloDBInternal.java [hunk 38]
```diff
@@ -73,7 +79,7 @@
 
     private HaloDBInternal() {}
 
-    static HaloDBInternal open(File directory, HaloDBOptions options) throws HaloDBException, IOException {
+    static synchronized HaloDBInternal open(File directory, HaloDBOptions options) throws HaloDBException, IOException {
```
Comment (wangtao724) [inline src/main/java/com/oath/halodb/HaloDBInternal.java]: Just noticed that this is a static method. ```synchronized``` will not do its work as what we expected
